### PR TITLE
CI: upgrade to setup-go v4 and disable cache with golangci

### PR DESCRIPTION
**Pull request by Shwetha-Acharya**
Below errors were seen with setup-go@v4 in golangci Lint: Error: /usr/bin/tar: ../../../go/pkg/mod/sigs.k8s.io/yaml@v1.3.0/yaml_ test.go: Cannot open: File exists
/usr/bin/tar: ../../../go/pkg/mod/sigs.k8s.io/yaml@v1.3.0/LICENSE: Cannot open: File exists

Disabling Cache with golangci solves the problem and makes sure that we are in a recent setup-go as well.

Cache can be enabled once the resultant issue is fixed in golangci lint action repo.
Issue that is to be tracked:
https://github.com/golangci/golangci-lint-action/issues/135

Updates: #1000

### CODE REVIEW

**Comment by Shwetha-Acharya** [comment]
Successful runs after the change:
https://github.com/Shwetha-Acharya/ramen/actions/runs/5666081903
https://github.com/RamenDR/ramen/actions/runs/5666374061/job/15352984165

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -58,6 +57,7 @@ jobs:
           git --no-pager diff
           git diff-index --quiet HEAD
 
+
```
Comment (nirs) [inline .github/workflows/ci.yaml:60]: The whitespace changes are unrelated and should be in another commit.

**Review by nirs — COMMENTED** [review]
It works, but I would check if this can be solved in golanci, since the issue is the cache used in golanci, not setup-go.

In any case, when we change the defaults we want to add a comment explaining why the change is needed.

**Review by ShyamsundarR — COMMENTED** [review]
The white space changes do not look consistent to me, not sure if the intent is to separate sections using 2 lines. Can you clarify @Shwetha-Acharya ?

**Comment by Shwetha-Acharya** [comment]
> It works, but I would check if this can be solved in golanci, since the issue is the cache used in golanci, not setup-go.
> 
> In any case, when we change the defaults we want to add a comment explaining why the change is needed.

As of now we are disabling cache on setup-go only when we are using golangci. 
I agree it would be better if there was a fix from golangci.

> The white space changes do not look consistent to me, not sure if the intent is to separate sections using 2 lines. Can you clarify @Shwetha-Acharya ?

I was experimenting with many changes to resolve this issue and did not notice that I had added extra spaces.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -66,15 +66,17 @@ jobs:
         uses: actions/checkout@v3
 
       - name: Setup go
-        uses: actions/setup-go@v3
+        uses: actions/setup-go@v4
         with:
           go-version: ${{ env.GO_VERSION }}
+          cache: false
```
[the remark this answers is not in the document]
Comment (Shwetha-Acharya) [reply .github/workflows/ci.yaml:77]: I agree, but I did not find any workarounds with golangci. 
I will raise a new issue in the golangci to highlight this scenario.
Also the golangci issue #135 has one partial fix which went in 2.4.1 and we are already in golangci-lint-action@v3.


I had tried by updating the' with version' in golangci/golangci-lint-action:
        uses: golangci/golangci-lint-action@v3
        with:
          version: v1.49.0
But did not find much luck with it. 


I am not sure what version v1.49.0 is actually representing.

[the remark this answers is not in the document]
Comment (Shwetha-Acharya) [reply .github/workflows/ci.yaml:77]: @nirs https://github.com/golangci/golangci-lint-action/issues/807

**Review by ShyamsundarR — APPROVED** [review]

**Review by nirs — APPROVED** [review]
